rye/src/cli/add.rs

```diff
@@ -21,12 +21,16 @@ from unearth.finder import PackageFinder
 
 py_ver = tuple(map(int, sys.argv[1].split('.')))
 package = sys.argv[2]
+if len(sys.argv) > 3 and sys.argv[3] == "--pre=True":
+    allow_prereleases = True
+else:
+    allow_prereleases = False
 
 finder = PackageFinder(
     index_urls=["https://pypi.org/simple/"],
 )
 finder.target_python.py_ver = py_ver
-m = finder.find_matches(package)
+m = finder.find_matches(package, allow_prereleases=allow_prereleases)
```

Comment:
Originally I used `packaging` but I didn't want to add that to the bootstrapping for `rye`. Ideally, this should work but it's not yet.

Reply:
IIUC the `allow_prereleases` param of `find_matches` is a condition to place on the supplied `requirement`

https://github.com/frostming/unearth/blob/ac8f934ff3f9316d9f76e4fd561ae878e5adc749/src/unearth/finder.py#L319C9-L325

```py
       # find_matches ...
        return LazySequence(
            self._evaluate_packages(
                self._find_packages_from_requirement(requirement, allow_yanked, hashes),
                requirement,
                allow_prereleases,
            )
        )
```

Reply:
IIUC the `allow_prereleases` param of `find_matches` is a condition to place on the *supplied* `requirement` instead of being used as a condition for a filtered `Iterable[Package]` to search.

Here's what `unearth` is doing with `allow_prereleases`:

1. Used to evaluate from found packages using `_evaluate_packages`
https://github.com/frostming/unearth/blob/ac8f934ff3f9316d9f76e4fd561ae878e5adc749/src/unearth/finder.py#L319C9-L325

```py
       # find_matches ...
        return LazySequence(
            self._evaluate_packages(
                self._find_packages_from_requirement(requirement, allow_yanked, hashes),
                requirement,
                allow_prereleases,
            )
        )
```

2. Passed to its `evaluator` on the supplied `requirement`.
https://github.com/frostming/unearth/blob/ac8f934ff3f9316d9f76e4fd561ae878e5adc749/src/unearth/finder.py#LL188C1-L199C43

```py
    def _evaluate_packages(
        self,
        packages: Iterable[Package],
        requirement: packaging.requirements.Requirement,
        allow_prereleases: bool | None = None,
    ) -> Iterable[Package]:
        evaluator = functools.partial(
            evaluate_package,
            requirement=requirement,
            allow_prereleases=allow_prereleases,
        )
        return filter(evaluator, packages)
```

3. If the supplied requirement contains a prerelease marker it'll skip.
https://github.com/frostming/unearth/blob/ac8f934ff3f9316d9f76e4fd561ae878e5adc749/src/unearth/evaluator.py#L285-L288

```py
   # evaluate_package ...
   if requirement.specifier and package.version:
        if not requirement.specifier.contains(
            package.version, prereleases=allow_prereleases
        ):
```

Since we're not using it this way, at the moment the workaround is all I have working. I can look into this more, but I'd rather understand where your mind is at with this one before more rabbit holes.

Reply:
fwiw [these changes](https://github.com/frostming/unearth/compare/main...cnpryer:unearth:allow-prerelease-find-packages) give me what I want. I'm thinking I'm just misunderstanding how to use `unearth`.

Reply:
I'm a bit puzzled by unearth's behavior here. I expected this to actually just work.